File: lib/core/Models/habit.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';

FirebaseFirestore firestore = FirebaseFirestore.instance;
var id = FirebaseAuth.instance.currentUser!.uid;

class Habit {
  String title;
  String description;
  DateTime startDate;
  DateTime endDate;
  int streak;
  bool isComplete;
  int totalCompletions;
  TimeOfDay reminderTime;
  List<Map<DateTime, bool>> days;
  List<String> selectedWeekdays;
  String userid = FirebaseAuth.instance.currentUser!.uid;
  String id;

  Habit({
    required this.title,
    required this.description,
    required this.startDate,
    required this.endDate,
    required this.streak,
    required this.isComplete,
    required this.totalCompletions,
    required this.reminderTime,
    required this.days,
    required this.selectedWeekdays,
    required this.userid,
    required this.id,
  });

  factory Habit.fromMap(Map<String, dynamic> data) {
    return Habit(
      id: data['id'] ?? '',
      title: data['title'],
      description: data['description'],
      startDate: data['startDate'].toDate(),
      endDate: data['endDate'].toDate(),
      streak: data['streak'],
      isComplete: data['isComplete'],
      totalCompletions: data['totalCompletions'],
      reminderTime: TimeOfDay(
        hour: data['reminderTime']['hour'],
        minute: data['reminderTime']['minute'],
      ),
      days: List<Map<DateTime, bool>>.from(data['days']),
      selectedWeekdays: List<String>.from(data['selectedWeekdays']),
      userid: data['userid'], 
    );
  }

  Future<void> addHabit() async {
    try {
      await firestore.collection('Habits').add({
        'title': title,
        'description': description,
        'startDate': Timestamp.fromDate(startDate),
        'endDate': Timestamp.fromDate(endDate),
        'streak': 0,
        'isComplete': false,
        'totalCompletions': totalCompletions,
        'reminderTime': {
          'hour': reminderTime.hour,
          'minute': reminderTime.minute,
        },
        'days': days,
        'selectedWeekdays': selectedWeekdays,
        'userid': userid,
      });
      print('Habit added for user: $userid');
    } catch (e) {
      print('Error adding habit: $e');
    }
  }

  Map<String, dynamic> toMap() {
    return {
      'title': title,
      'description': description,
      'startDate': Timestamp.fromDate(startDate),
      'endDate': Timestamp.fromDate(endDate),
      'streak': streak,
      'isComplete': isComplete,
      'totalCompletions': totalCompletions,
      'reminderTime': {
        'hour': reminderTime.hour,
        'minute': reminderTime.minute,
      },
      'days': days,
      'selectedWeekdays': selectedWeekdays,
      'userid': userid,
    };
  }
}

File: Backend/src/model/habit.js
import { DynamoDBClient, PutItemCommand } from "@aws-sdk/client-dynamodb";
import { formatISO, parseISO } from "date-fns";

const client = new DynamoDBClient({ region: "ap-south-1" });

class Habit {
  constructor({
    title,
    description,
    startDate,
    endDate,
    streak,
    isComplete,
    totalCompletions,
    reminderTime,
    days,
    selectedWeekdays,
    userid,
    id,
  }) {
    this.title = title;
    this.description = description;
    this.startDate = startDate;
    this.endDate = endDate;
    this.streak = streak;
    this.isComplete = isComplete;
    this.totalCompletions = totalCompletions;
    this.reminderTime = reminderTime; 
    this.days = days; 
    this.userid = userid;
    this.id = id;
  }

  toMap() {
    return {
      title: { S: this.title },
      description: { S: this.description },
      startDate: { S: formatISO(this.startDate) },
      endDate: { S: formatISO(this.endDate) },
      streak: { N: this.streak.toString() },
      isComplete: { BOOL: this.isComplete },
      totalCompletions: { N: this.totalCompletions.toString() },
      reminderTime: {
        M: {
          hour: { N: this.reminderTime.hour.toString() },
          minute: { N: this.reminderTime.minute.toString() },
        },
      },
      days: {
        L: this.days.map((day) => ({
          M: {
            date: { S: formatISO(day.date) },
            isComplete: { BOOL: day.isComplete },
          },
        })),
      },
      selectedWeekdays: { L: this.selectedWeekdays.map((day) => ({ S: day })) },
      userid: { S: this.userid },
      id: { S: this.id },
    };
  }

  static fromMap(map) {
    return new Habit({
      title: map.title.S,
      description: map.description.S,
      startDate: parseISO(map.startDate.S),
      endDate: parseISO(map.endDate.S),
      streak: parseInt(map.streak.N, 10),
      isComplete: map.isComplete.BOOL,
      totalCompletions: parseInt(map.totalCompletions.N, 10),
      reminderTime: {
        hour: parseInt(map.reminderTime.M.hour.N, 10),
        minute: parseInt(map.reminderTime.M.minute.N, 10),
      },
      days: map.days.L.map((day) => ({
        date: parseISO(day.M.date.S),
        isComplete: day.M.isComplete.BOOL,
      })),
      selectedWeekdays: map.selectedWeekdays.L.map((day) => day.S),
      userid: map.userid.S,
      id: map.id.S,
    });
  }

  async save() {
    const params = {
      TableName: "Habits",
      Item: this.toMap(),
    };
    await client.send(new PutItemCommand(params));
  }
}

export default Habit;